DM: update openapi spec (#7866)
pingcap/docs: dm/dm-open-api.md

# Maintain DM Clusters Using OpenAPI

DM provides the OpenAPI feature for easily querying and operating the DM cluster, which is similar to the feature of [dmctl tools](/dm/dmctl-introduction.md).

To enable OpenAPI, perform one of the following operations:

+ If your DM cluster has been deployed directly using binary, add the following configuration to the DM-master configuration file.

    ```toml
    openapi = true
    ```

> + If your DM cluster is deployed using TiUP, add the following configuration to the topology file：

Reviewer: ```suggestion
+ If your DM cluster has been deployed using TiUP, add the following configuration to the topology file：
```

    {
      "source_name": "mysql-replica-01",
      "worker_name": "worker-1",
      "relay_status": {
        "master_binlog": "(mysql-bin.000001, 1979)",
        "master_binlog_gtid": "e9a1fc22-ec08-11e9-b2ac-0242ac110003:1-7849",
        "relay_dir": "./sub_dir",
        "relay_binlog_gtid": "e9a1fc22-ec08-11e9-b2ac-0242ac110003:1-7849",
        "relay_catch_up_master": true,
        "stage": "Running"
      },
      "error_msg": "string"
    }
  ],
  "relay_config": {
    "enable_relay": true,
    "relay_binlog_name": "mysql-bin.000002",
    "relay_binlog_gtid": "e9a1fc22-ec08-11e9-b2ac-0242ac110003:1-7849",
    "relay_dir": "./relay_log"
  }
}
```

## Delete the data source

This API is a synchronous interface. If the request is successful, the status code of the returned body is 204.

### Request URI

`DELETE /api/v1/sources/{source-name}`

### Example

{{< copyable "shell-regular" >}}

```shell
curl -X 'DELETE' \
  'http://127.0.0.1:8261/api/v1/sources/mysql-01?force=true' \
  -H 'accept: application/json'
```

## Update a data source

This API is a synchronous interface. If the request is successful, the information of the corresponding data source is returned.

> > **NOTE:**
> >
> > Updating the data source configuration requires that there are no running tasks under the current data source

Reviewer: ```suggestion

> **NOTE:**
>
> When you use this API to update the data source configuration, make sure that there are no running tasks under the current data source.
```

  -H 'Content-Type: application/json' \
  -d '{
  "source": {
    "source_name": "mysql-01",
    "host": "127.0.0.1",
    "port": 3306,
    "user": "root",
    "password": "123456",
    "enable_gtid": false,
    "enable": false,
    "flavor": "mysql",
    "task_name_list": [
      "task1"
    ],
    "security": {
      "ssl_ca_content": "",
      "ssl_cert_content": "",
      "ssl_key_content": "",
      "cert_allowed_cn": [
        "string"
      ]
    },
    "purge": {
      "interval": 3600,
      "expires": 0,
      "remain_space": 15
    },
    "relay_config": {
      "enable_relay": true,
      "relay_binlog_name": "mysql-bin.000002",
      "relay_binlog_gtid": "e9a1fc22-ec08-11e9-b2ac-0242ac110003:1-7849",
      "relay_dir": "./relay_log"
    }
  }
}'
```

```json
{
  "source_name": "mysql-01",
  "host": "127.0.0.1",
  "port": 3306,
  "user": "root",
  "password": "123456",
  "enable": true,
  "enable_gtid": false,
  "security": {
    "ssl_ca_content": "",
    "ssl_cert_content": "",
    "ssl_key_content": "",
    "cert_allowed_cn": [
      "string"
    ]
  },
  "purge": {
    "interval": 3600,
    "expires": 0,
    "remain_space": 15
  }
}
```

> ## Enable-a data source

Reviewer: ```suggestion
## Enable a data source
```

> This is a synchronisation interface that enables this data source on a successful request and starts all subtasks of the task that rely on this data source in bulk.

Reviewer: ```suggestion
This is a synchronous interface that enables a data source on a successful request and starts all subtasks of the task that rely on this data source in batch.
```

### Request URI

`POST /api/v1/sources/{source-name}/enable`

### Example

{{< copyable "shell-regular" >}}

```shell
curl -X 'POST' \
  'http://127.0.0.1:8261/api/v1/sources/mysql-01/enable' \
  -H 'accept: */*' \
  -H 'Content-Type: application/json'
```

## Disable a data source

> This is a synchronisation interface that deactivates this data source on a successful request and stops all subtasks of the task that depend on it in bulk.

Reviewer: ```suggestion
This is a synchronous interface that deactivates this data source on a successful request and stops all subtasks of the task that rely on it in batch.
```

}'
```

> ## Purge relay log files that no longer required

Reviewer: ```suggestion
## Purge relay log files that are no longer required
```

## Create a replication task

> This API is an synchronous interface. If the request is successful, the status code of the returned body is 200. a successful request will return information about the corresponding synchronisation task.

Reviewer: ```suggestion
This API is a synchronous interface. If the request is successful, the status code of the returned body is 200. A successful request will return the information of the corresponding replication task.
```

        "source_name": "source-name",
        "schema": "db-*",
        "table": "tb-*"
      },
      "target": {
        "schema": "db1",
        "table": "tb1"
      },
      "binlog_filter_rule": [
        "rule-1",
        "rule-2",
        "rule-3",
      ]
    }
  ],
  "source_config": {
    "full_migrate_conf": {
      "export_threads": 4,
      "import_threads": 16,
      "data_dir": "./exported_data",
      "consistency": "auto"
    },
    "incr_migrate_conf": {
      "repl_threads": 16,
      "repl_batch": 100
    },
    "source_conf": [
      {
        "source_name": "mysql-replica-01",
        "binlog_name": "binlog.000001",
        "binlog_pos": 4,
        "binlog_gtid": "03fc0263-28c7-11e7-a653-6c0b84d59f30:1-7041423,05474d3c-28c7-11e7-8352-203db246dd3d:1-170"
      }
    ]
  }
}
```

## Delete a replication task

This interface is a synchronous interface and the Status Code of the returned body is 204 upon successful request.

### Request URI

`DELETE /api/v1/tasks/{task-name}`

### Example

{{< copyable "shell-regular" >}}

```shell
curl -X 'DELETE' \
  'http://127.0.0.1:8261/api/v1/tasks/task-1' \
  -H 'accept: application/json'
```

## Update a replication task

> This interface is a synchronisation interface and a successful request will return information about the task.

Reviewer: ```suggestion
This interface is a synchronous interface and a successful request returns the information of the task.
```

> > **NOTE:**
> >
> > Updating the task configuration requires that the task is stopped and has run into incremental sync and that only some of the fields can be updated.

Reviewer: ```suggestion
> **Note:**
>
> When you use this API to update the task configuration, make sure that the task is stopped and has run into incremental sync and that only some of the fields can be updated.
```

      ],
      "ignore_sql": [
        "^Drop"
      ]
    },
    "rule-2": {
      "ignore_event": [
        "all dml"
      ],
      "ignore_sql": [
        "^Drop"
      ]
    },
    "rule-3": {
      "ignore_event": [
        "all dml"
      ],
      "ignore_sql": [
        "^Drop"
      ]
    }
  },
  "table_migrate_rule": [
    {
      "source": {
        "source_name": "source-name",
        "schema": "db-*",
        "table": "tb-*"
      },
      "target": {
        "schema": "db1",
        "table": "tb1"
      },
      "binlog_filter_rule": [
        "rule-1",
        "rule-2",
        "rule-3",
      ]
    }
  ],
  "source_config": {
    "full_migrate_conf": {
      "export_threads": 4,
      "import_threads": 16,
      "data_dir": "./exported_data",
      "consistency": "auto"
    },
    "incr_migrate_conf": {
      "repl_threads": 16,
      "repl_batch": 100
    },
    "source_conf": [
      {
        "source_name": "mysql-replica-01",
        "binlog_name": "binlog.000001",
        "binlog_pos": 4,
        "binlog_gtid": "03fc0263-28c7-11e7-a653-6c0b84d59f30:1-7041423,05474d3c-28c7-11e7-8352-203db246dd3d:1-170"
      }
    ]
  }
}
```

## Start a replication task

> This API is an asynchronous interface. If the request is successful, the status code of the returned body is 204. To learn about its latest status, You can [get the information of a replication task](#get-the-information-of-a-replication-task).

Reviewer: ```suggestion
This API is an asynchronous interface. If the request is successful, the status code of the returned body is 204. To learn the latest status of a task, You can [get the information of a replication task](#get-the-information-of-a-replication-task).
```

### Request URI

`POST /api/v1/tasks/{task-name}/start`

### Example

{{< copyable "shell-regular" >}}

```shell
curl -X 'POST' \
  'http://127.0.0.1:8261/api/v1/tasks/task-1/start' \
  -H 'accept: */*'
```

## Stop a replication task

> This API is an asynchronous interface. If the request is successful, the status code of the returned body is 204. To learn about its latest status, You can [get the information of a replication task](#get-the-information-of-a-replication-task).

Reviewer: ```suggestion
This API is an asynchronous interface. If the request is successful, the status code of the returned body is 200. To learn the latest status of a task, You can [get the information of a replication task](#get-the-information-of-a-replication-task).
```

}
```

## Get the replication task list

> This API is a synchronous interface. If the request is successful, the information of the corresponding node is returned.

Reviewer: 这里中文是【请求成功会返回对应的同步任务信息】，似乎不太对应呢：https://github.com/pingcap/docs-cn/pull/8612/files#diff-2d25f555740c97a16d90319fb4179091a04117e4ab95354d871cf2712c04e85eR1344

Author: 这里之前的中文描述就不太准确，感谢指正，在这里修复了

https://github.com/pingcap/docs-cn/pull/8612/commits/eaac731c2700aaff6c007a69ede874291624d071
https://github.com/pingcap/docs/pull/7866/commits/2da947d5f1d1f9dd0a80902aaf278be118b5e8ef

        },
        "rule-3": {
          "ignore_event": [
            "all dml"
          ],
          "ignore_sql": [
            "^Drop"
          ]
        }
      },
      "table_migrate_rule": [
        {
          "source": {
            "source_name": "source-name",
            "schema": "db-*",
            "table": "tb-*"
          },
          "target": {
            "schema": "db1",
            "table": "tb1"
          },
          "binlog_filter_rule": [
            "rule-1",
            "rule-2",
            "rule-3",
          ]
        }
      ],
      "source_config": {
        "full_migrate_conf": {
          "export_threads": 4,
          "import_threads": 16,
          "data_dir": "./exported_data",
          "consistency": "auto"
        },
        "incr_migrate_conf": {
          "repl_threads": 16,
          "repl_batch": 100
        },
        "source_conf": [
          {
            "source_name": "mysql-replica-01",
            "binlog_name": "binlog.000001",
            "binlog_pos": 4,
            "binlog_gtid": "03fc0263-28c7-11e7-a653-6c0b84d59f30:1-7041423,05474d3c-28c7-11e7-8352-203db246dd3d:1-170"
          }
        ]
      }
    }
  ]
}
```

> ## Get the replication task migrate rules

Reviewer: ```suggestion
## Get the migration rules of a replication task
```

> This API is a synchronous interface and a successful request will return the corresponding list.

Reviewer: ```suggestion
This API is a synchronous interface and a successful request returns a list of the corresponding tasks.
```

> This API is a synchronous interface and a successful request will return the corresponding list.

Reviewer: ```suggestion
This API is a synchronous interface and a successful request returns a list of the corresponding tasks.
```

            "all dml"
          ],
          "ignore_sql": [
            "^Drop"
          ]
        }
      },
      "table_migrate_rule": [
        {
          "source": {
            "source_name": "source-name",
            "schema": "db-*",
            "table": "tb-*"
          },
          "target": {
            "schema": "db1",
            "table": "tb1"
          },
          "binlog_filter_rule": [
            "rule-1",
            "rule-2",
            "rule-3",
          ]
        }
      ],
      "source_config": {
        "full_migrate_conf": {
          "export_threads": 4,
          "import_threads": 16,
          "data_dir": "./exported_data",
          "consistency": "auto"
        },
        "incr_migrate_conf": {
          "repl_threads": 16,
          "repl_batch": 100
        },
        "source_conf": [
          {
            "source_name": "mysql-replica-01",
            "binlog_name": "binlog.000001",
            "binlog_pos": 4,
            "binlog_gtid": "03fc0263-28c7-11e7-a653-6c0b84d59f30:1-7041423,05474d3c-28c7-11e7-8352-203db246dd3d:1-170"
          }
        ]
      }
    }
  ]
}
```

## Get the migration rules of a replication task

> This API is a synchronous interface and a successful request returns a list of the migrate rules of this task.

Reviewer: ```suggestion
This API is a synchronous interface and a successful request returns a list of the migration rules of this task.
```
